test/Data/Source/Spec.hs

```diff
@@ -1,53 +1,77 @@
+
+    , prop "computes sub-line ranges" $
+        \ s -> let source = "*" <> s <> "*" in
+          spanToRange source (insetSpan (totalSpan source)) === insetRange (totalRange source)
+
+    , testProperty "inverse of rangeToSpan" . property $ do
+        a <- forAll . Gen.source $ Hedgehog.Range.linear 0 100
+        b <- forAll . Gen.source $ Hedgehog.Range.linear 0 100
+        let s = a <> "\n" <> b in spanToRange s (totalSpan s) === totalRange s
+    ]
+
+  ,  testProperty "rangeToSpan inverse of spanToRange" . property $ do
+      a <- forAll . Gen.source $ Hedgehog.Range.linear 0 100
+      b <- forAll . Gen.source $ Hedgehog.Range.linear 0 100
+      let s = a <> "\n" <> b in rangeToSpan s (totalRange s) === totalSpan s
+
+  , Tasty.testGroup "totalSpan"
+    [ testProperty "covers single lines" . property $ do
+        n <- forAll $ Gen.int (Hedgehog.Range.linear 0 100)
+        totalSpan (fromText (Text.replicate n "*")) === Span (Pos 1 1) (Pos 1 (max 1 (succ n)))
+
+    , testProperty "covers multiple lines" . property $ do
+        n <- forAll $ Gen.int (Hedgehog.Range.linear 0 100)
+        totalSpan (fromText (Text.intersperse '\n' (Text.replicate n "*"))) === Span (Pos 1 1) (Pos (max 1 n) (if n > 0 then 2 else 1))
+    ]
+
+  ]
 
 spec :: Spec
 spec = parallel $ do
-  describe "sourceLineRanges" $ do
-    prop "produces 1 more range than there are newlines" $
-      \ source -> length (sourceLineRanges source) `shouldBe` succ (Text.count "\n" (toText source))
-
-    prop "produces exhaustive ranges" $
-      \ source -> foldMap (`slice` source) (sourceLineRanges source) `shouldBe` source
-
-  describe "spanToRange" $ do
-    prop "computes single-line ranges" $
-      \ s -> let source = fromUTF8 s
-                 spans = zipWith (\ i Range {..} -> Span (Pos i 1) (Pos i (succ (end - start)))) [1..] ranges
-                 ranges = sourceLineRanges source in
-        spanToRange source <$> spans `shouldBe` ranges
-
-    prop "computes multi-line ranges" $
-      \ source ->
-        spanToRange source (totalSpan source) `shouldBe` totalRange source
-
-    prop "computes sub-line ranges" $
-      \ s -> let source = "*" <> s <> "*" in
-        spanToRange source (insetSpan (totalSpan source)) `shouldBe` insetRange (totalRange source)
-
-    prop "inverse of rangeToSpan" $
-      \ a b -> let s = a <> "\n" <> b in spanToRange s (totalSpan s) `shouldBe` totalRange s
-
-  describe "rangeToSpan" $ do
-    prop "inverse of spanToRange" $
-      \ a b -> let s = a <> "\n" <> b in rangeToSpan s (totalRange s) `shouldBe` totalSpan s
-
-  describe "totalSpan" $ do
-    prop "covers single lines" $
-      \ n -> totalSpan (fromText (Text.replicate n "*")) `shouldBe` Span (Pos 1 1) (Pos 1 (max 1 (succ n)))
-
-    prop "covers multiple lines" $
-      \ n -> totalSpan (fromText (Text.intersperse '\n' (Text.replicate n "*"))) `shouldBe` Span (Pos 1 1) (Pos (max 1 n) (if n > 0 then 2 else 1))
-
   describe "newlineIndices" $ do
```

Comment:
In keeping with my philosophy of “port as little as possible over”, I’ve left these non-property tests as specs.